File: CPP/No684.cc
/**
 * Created by Xiaozhong on 2020/11/16.
 * Copyright (c) 2020/11/16 Xiaozhong. All rights reserved.
 */
#include <vector>
#include <iostream>
#include <unordered_set>
#include <unordered_map>

using namespace std;

class Solution {
    unordered_map<int, unordered_set<int>> adjacent;

    bool check(const int from, const int prev, const int to) {
        if (from == to) return true;
        auto dists = adjacent[from];
        for (auto dist : dists) {
            if (dist == prev) continue;
            if (check(dist, from, to)) return true;
        }
        return false;
    }

public:
    vector<int> findRedundantConnection(vector<vector<int>> &edges) {
        int n = edges.size();
        for (const auto &edge : edges) {
            bool res = check(edge.front(), 0, edge.back());
            if (res) return edge;
            else {
                int from = edge.front(), to = edge.back();
                adjacent[from].insert(to);
                adjacent[to].insert(from);
            }
        }
        return {};
    }
};

int main() {
    Solution s;
    vector<vector<int>> edges = {
            {1, 2},
            {1, 3},
            {2, 3}
    };
    vector<int> ans = s.findRedundantConnection(edges);
    return 0;
}

File: Java/src/No684.java
public class No684 {
    // 准备返回值
    private int[] ans = new int[2];
    // 用来标记连通 identifier 的数组
    private int[] id;
    // 题目提供的接口，给我们一个 edges 数组，用来表示边。
    public int[] findRedundantConnection(int[][] edges) {
        // 为什么是 n + 1？还不是因为题目没有从 0 开始计数😑
        id = new int[edges.length + 1];
        // 初始化各个边
        for (int i = 1; i < id.length; ++i) id[i] = i;
        // 把各个边都添加进连通分量中去
        for (int[] edge : edges) union(edge[0], edge[1]);
        return ans;
    }

    private void union(int p, int q) {
        int pRoot = find(p);
        int qRoot = find(q);
        // p 和 q 并没有连通，那么就让他们连通
        if (pRoot != qRoot) id[pRoot] = qRoot;
        // 说明 p 和 q 已经连通了呀，那么就直接把它们记录下来
        else {
            ans[0] = p;
            ans[1] = q;
        }
    }

    private int find(int p) {
        // 向上回查根节点编号
        while (id[p] != p) p = id[p];
        return p;
    }

    public static void main(String[] args) {

    }
}
